p2panda-stream/src/ordering/partial/mod.rs

```diff
@@ -0,0 +1,368 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+pub mod operations;
+pub mod store;
+
+use std::collections::VecDeque;
+use std::fmt::{Debug, Display};
+use std::hash::Hash as StdHash;
+
+use thiserror::Error;
+
+use store::PartialOrderStore;
+
+/// Error types which may be returned from `PartialOrder` methods.
+#[derive(Debug, Error)]
+pub enum PartialOrderError {
+    #[error("store error: {0}")]
+    StoreError(String),
+}
+
+/// Struct for establishing partial order over items in a Directed-Acyclic-Graph.
+///
+/// There are various approaches which can be taken when wanting to linearize items in a graph
+/// structure. This approach establishes a partial order, meaning not all items in the graph are
+/// comparable, and is non-deterministic. The main requirement is that all dependencies of an item
+/// are sorted "before" the item itself, the exact order is not a concern.
+///
+/// Items are considered in a "ready" state when all dependencies are met, and in a "pending"
+/// state when their dependencies are not yet met. Calls to `next` will return "ready" items in
+/// the order that they became ready.
+///
+/// If an item is processed an in a "pending" state then it is held in a pending queue and if it's
+/// dependencies are later processed and "ready", then the so far "pending" item will be moved to
+/// the "ready" queue. This processing of pending items recursively checks all pending dependents.
+///
+/// Example graph:
+///
+/// A <-- B2 <-- C
+///   \-- B1 <--/
+///
+/// Both of the following are possible and valid orderings for the above graph:
+///
+/// [A, B1, B2, C]
+/// [A, B2, B1, C]
+///
+/// Items will not be placed into an partial order until all their dependencies are met, in the
+/// following example item C will not be visited as we have not processed all of it's
+/// dependencies.
+///
+/// Example graph:
+///
+/// A <-- ?? <-- C
+///   \-- B1 <--/
+///
+/// C is not processed yet as we are missing one of its dependencies:
+///
+/// [A, B1]
+///
+/// Note that no checks are made for cycles occurring in the graph, this should be validated on
+/// another layer.
+#[derive(Debug)]
+pub struct PartialOrder<K, S> {
+    /// Store for managing "ready" and "pending" items.
+    store: S,
+
+    /// Sorted queue of "ready" items.
+    ready_queue: VecDeque<K>,
+}
+
+impl<K, S> PartialOrder<K, S>
+where
+    K: Clone + Copy + Display + StdHash + PartialEq + Eq,
+    S: PartialOrderStore<K>,
+{
+    pub fn new(store: S) -> Self {
+        Self {
+            store,
+            ready_queue: VecDeque::new(),
+        }
+    }
+
+    /// Pop the next item from the ready queue.
+    pub fn next(&mut self) -> Option<K> {
+        self.ready_queue.pop_front()
```

Comment:
Is it possible to mark an item as "ready" _after_ it got popped from the queue?

I'm thinking about crash resiliance here as we might get intro trouble when items got "freed" by being marked as ready, _then_ moved into the ready queue. IF the process crashes we would loose them and that information as the ready queue only lives in memory.

I can see that it would be easy to just mark them as ready as soon as they get returned, but maybe that messes with the logic in other parts?

Reply:
In a "robust" implementation which would use this I can see how one would have a single SQL transaction for whatever the application does next and marking this as "ready", to not loose any data in that transition.